ntoskrnl/mm/ARM3/zeropage.c

```diff
@@ -73,7 +72,7 @@ MmZeroPageThread(VOID)
         {
             if (!MmFreePageListHead.Total)
             {
-                MmZeroingPageThreadActive = FALSE;
+                KeClearEvent(&MmZeroingPageEvent);
```

Comment:
Is it possible that a thread calls `MiInsertPageInFreeList()` and manages to set the event again (see your modification in the `if (ListHead->Total >= 8)` block of that function in pfnlist.c) while the execution of this `MmZeroPageThread()` thread happens to be in this `if (!MmFreePageListHead.Total)` block (i.e. past the `KeWaitForMultipleObjects()` call but just before clearing out the event) ? Because if so this would signify that we could miss the fact that `MiInsertPageInFreeList()` set the event and then in the zeroing thread we would go back to waiting for the event to be set...

Reply:
Both that `KeSetEvent` and this `KeClearEvent` hold the PFN lock, so it should be sufficiently guarded (the same way that the boolean was)

Reply:
You mean that they are called under PFN lock acquired? hmm ok....